### ISSUE TITLE
Stutters after cursor auto hide in every game.

### ISSUE THREAD

**Issue by Chococococococo** [ValveSoftware/gamescope#1056]
The game stutters every time the mouse cursor disappear in steam deck game mode. It happens in every game. Not sure what cause this. The frame drop can be clearly see in gamescope overlay fps drop and frametime spike.

The only way to avoid is not to use any mouse input. I tried to change every setting mouse related but nothing help.

**Comment by BeppisMAX** [comment]
Some notes regarding this issue:

- On the Steam Deck, this is only apparent when running games in Game Mode, where Gamescope is utilized.
- Appears to only happen when a hardware-based mouse cursor is being used in-game. Some games will use this by default, or will allow configuration to use either a hardware or software-based cursor in-game.
- As @david-rosca mentioned, eliminating the occurance will require FSR upscaling to be in-effect. As soon as FSR is disabled however, the issue will persist.
- On SteamOS, this issue seems to have been apparent since the 3.5 release. Changing the version of Proton in which a game runs does not appear to change the behavior.
- The lower the framerate, the lower FPS the stutter will cause the game to drop to.

Testing by the community has been conducted on several games. On my end, I've been able reliabily duplicate the issue on Final Fantasy XIV. Steps to duplicate for this particular game below.

1. On the Steam Deck, open the game in Game Mode with FSR upscaling off. (Optional: Enable the in-built Performance overlay in order to track FPS)
2. Log in via the launcher.
3. On the splash/main menu, navigate to Configuration or log into a Character, then access the Main Menu, navigate to System > System Configuration.
4. Navigate to Mouse setting, and ensure Hardware cursor is enabled.
5. Ensure your Steam controller configuration allows for mouse input.
6. Triggering the mouse to 'hide' either via click-and-hold to move the camera or allowing a few seconds of idle activity to hide the cursor, will trigger the stutter - causing the FPS to noticably tank.

The circumstances in which a cursor will 'hide' for each game my vary, however it seems when running a game in Gamescope, a hardware-based cursor entering a hidden state will 'tank' the FPS briefly. The severity of the issue could vary on how a game may rely on the appearance and disappearance of a mouse cursor. A hardware-based cursor serves to be a preference among users due to the immediate response rate and system's mouse configuration.

**Comment by sharkautarch** [comment]
@Joshua-Ashton 
Unless you had already figured out what was causing this issue, 
I think the best way of finding the root cause of this issue would be to hook up [Tracy profiler](https://github.com/wolfpld/tracy) to gamescope.
It tracks start and end times of frame prep/submission like w/ gpuvis, but also helps you track down the problem code that could be delaying frame submission at any given point in time.

**Comment by david-rosca** [comment]
This same issue seems to be present with Homeworld 3 as well.

**Comment by Eduardboon** [comment]
Whenever the mouse cursor shows on an embedded gamescope session (like gaming mode on the steam deck) it forces composite mode instead of scanout. This makes the whole screen slightly blurry as well. 

I think the switching of these modes creates the fps spike as well, since the rendering mode changes. 

I’ve been trying to fix the blurry part of the issue without any success at all. Since in most games that use the cursor I now have to play the entire game blurred. Or when the cursor disappears, enjoy the screen constantly switching between sharp and blurry rendering modes.

**Comment by trevyn** [comment]
Got hit by this as well; writing a little game experiment, and I'm like why is this stuttering after I move the mouse? I think @Eduardboon nailed what's going on. I wonder if there's a way to composite just a hardware cursor without going into full compositor mode.

**Comment by sharkautarch** [comment]
@trevyn 
Hmm my one idea for "drawing" a cursor without doing full composition (switching to/from blur layer) would be:
- Make use of `VK_KHR_incremental_present` (And might also need to use `VK_KHR_display` to present onto the drm surface) to vulkan-composite a small rectangle that fits the shape of the cursor (and probably also need to composite another small rectangle over where the cursor was at in the previous frame)

**Comment by Joshua-Ashton** [comment]
No, that's not how this works at all.

We should just use the cursor plane but AMDGPU bugs prevent us right now. I'll try and look at it soon.

**Comment by Joshua-Ashton** [comment]
(FWIW we also have partial composition mostly implememted. I forget the reason that it is disabled right now)

**Comment by Eduardboon** [comment]
So is this normal compositing behaviour then? ([Link to video](https://drive.google.com/file/d/1u7DP1JIy6Yqr3dGI9XDnQrJHnvejDgFj/view?usp=drivesdk)

Because I could have sworn it didn’t blur by this much when using the cursor

**Comment by trevyn** [comment]
I was also able to reliably reproduce the blur/no-blur effect on Steam Deck in any game by setting a higher-than-native resolution in the Steam display properties, e.g. 1920x1080.

**Comment by Eduardboon** [comment]
I am getting the blur even on the native res of 1280x800 at any refresh rate. It doesn’t matter what resolution I use, blur always shows up here. 

My gamescope log has started spitting out this line a lot:
![IMG_2343](https://github.com/user-attachments/assets/841e962a-5f52-4edb-8123-197c551a4b86)

**Comment by ubeogesh** [comment]
This issue is very annoying in Diablo 2 resurrected, because when you pick up and put down items, game replaces cursor with the item graphic - and backwards - which causes visible stutters.

**Comment by ubeogesh** [comment]
Hi yall,

I wonder if it's possible ro force FSR on somehow to work around this bug, when the game I'm playing doesn't support a resolution below 1280x800? Somehow trick the FSR filter to think the screen resolution is, for example, 1600x1000?

**Comment by sharkautarch** [comment]
@ubeogesh @Chococococococo 
I had a discussion with @misyltoad & @matte-schwartz about a group of issues wrt stuttering happening only on game mode.
Tho, since people in this issue thread were talking about FSR mitigating the issue, I'm not entirely sure whether this is a separate issue from the one I had discussed about elsewhere.
Anyhow, to summarize, I had reviewed gamescope's compositing code -- specifically the logic for selecting which layers to paint -- and came to the realization that theoretically, if multiple unrelated events (that all increment the same shared internal counter in gamescope) occurred in a short span of time, gamescope would be forced to start flipping a frame which doesn't include the main-layer/window.
To my knowledge, this is still only a conjecture, and hasn't actually been confirmed to be the root cause of any stutter-related issues yet.
Btw, the specific aforementioned internal counter is ` nIgnoredOverlayRepaints`
And you can use `gamescopectl` to see if changing the number of ignored repaint requests before repaint is forced, or to ignore all overlay repaint requests with the `adaptive_sync_overlay_cycles` and `adaptive_sync_ignore_overlay` `gamescopectl` variables respectively.

**Comment by fdesuhijfsbnijukfdsjkigfv** [comment]
Can confirm this is still an issue on the latest version of stable SteamOS. Has anyone discovered a workaround?

**Comment by Regnuss** [comment]
> Can confirm this is still an issue on the latest version of stable SteamOS. Has anyone discovered a workaround?

Yeah still having this issue on the Steam Deck OLED, only fix is to disable hardware cursor (enable software cursor) in games that allow it or play in desktop mode. Enabling FSR does not fix it, it still stutters everytime the game hides the mouse if you use the hardware cursor in games like WoW (classic).

**Comment by ubeogesh** [comment]
FSR workaround stopped working in 3.7. From the frametime graph i see stutters when the cursor hides - not when it's shown. 

![Image](https://github.com/user-attachments/assets/503300c6-33f7-43d6-a26a-e6c5c74d90c9)
![Image](https://github.com/user-attachments/assets/b20abfc5-bd5d-4aa2-85e7-99c4804ef23d)

**Comment by Bradaloop** [comment]
FYI that this is still an issue.

**Comment by Adam-Beik** [comment]
Still an issue. Experiencing this in Guild Wars Reforged when moving camera with right-click held while playing in game mode.

**Comment by OursCodeur** [comment]
Did some instrumented digging on the embedded DRM backend:

- The cursor-show stutter is deterministic and not AMD-specific: with the cursor plane hardcoded off, a visible cursor forces full composition, and the **first composited frame missed a vblank on 15/15 cursor shows** (33 ms frame delta). Two stacked causes: the vblank timer only learns about compositing one frame late, so the first composite gets a scanout-sized (~7 ms) wakeup budget while needing ~11 ms; and the first compose on an idle GPU runs before clocks ramp (15.6 ms vs 8.2 ms steady here).
- Cursor **hide** costs nothing extra on i915, so the hide-side spikes on Deck are consistent with the AMDGPU plane/color-mgmt transition cost @misyltoad mentioned, stacked on the same toggle.
- Enabling the cursor plane removes the toggle entirely: 30+ show/hide cycles with direct scanout throughout, p99 frame delta 16.91 ms, zero drops, cursor renders correctly (i915). It also avoids the composite scaling blur while the cursor is visible that was reported later in this thread.

Opened https://github.com/ValveSoftware/gamescope/pull/2316 putting the cursor plane behind a `drm_cursor_plane` convar (default off, so no behavior change) so this is testable on Deck/AMDGPU to see where the driver issues stand today.
